datatypes/src/collections/feature_collection.rs

```diff
@@ -766,11 +843,101 @@ fn struct_array_from_data(
     )
 }
 
+/// Types that are suitable to act as filters
+pub trait FilterArray: Into<BooleanArray> {
+    fn len(&self) -> usize;
+}
+
+impl FilterArray for Vec<bool> {
+    fn len(&self) -> usize {
+        Vec::<_>::len(self)
+    }
+}
+
+impl FilterArray for BooleanArray {
+    fn len(&self) -> usize {
+        <Self as arrow::array::Array>::len(self)
+    }
+}
+
+fn update_filter_array(
+    filter_array: &mut Option<BooleanArray>,
+    partial_filter_a: Option<BooleanArray>,
+    partial_filter_b: Option<BooleanArray>,
+) -> Result<()> {
+    let partial_filter = match (partial_filter_a, partial_filter_b) {
+        (Some(f1), Some(f2)) => Some(arrow::compute::and(&f1, &f2)?),
+        (Some(f1), None) => Some(f1),
+        (None, Some(f2)) => Some(f2),
+        (None, None) => None,
+    };
+
+    *filter_array = match (filter_array.take(), partial_filter) {
+        (Some(f1), Some(f2)) => Some(arrow::compute::or(&f1, &f2)?),
+        (Some(f1), None) => Some(f1),
+        (None, Some(f2)) => Some(f2),
+        (None, None) => None,
+    };
+
+    Ok(())
+}
+
+fn apply_filter_on_bound<'b, T, A>(
+    bound: Bound<&'b FeatureDataValue>,
+    array: &'b A,
+    included_fn: fn(&'b A, T) -> Result<BooleanArray, ArrowError>,
+    excluded_fn: fn(&'b A, T) -> Result<BooleanArray, ArrowError>,
+) -> Result<Option<BooleanArray>>
+where
+    T: TryFrom<&'b FeatureDataValue, Error = error::FeatureCollectionError>,
+{
+    Ok(match bound {
+        Bound::Included(v) => Some(included_fn(array, v.try_into()?)?),
```

Comment:
why differentiate between bound types? Don't bounds provide a contains method that abstracts this away?